Pcap++/src/PcapLiveDevice.cpp

```diff
@@ -1056,33 +1063,118 @@ void PcapLiveDevice::setDefaultGateway()
 			PCPP_LOG_ERROR("Error retrieving default gateway address: " << e.what());
 		}
 	}
-#elif defined(__APPLE__) || defined(__FreeBSD__)
-	std::string command = "netstat -nr | grep default | grep " + m_Name;
-	std::string ifaceInfo = executeShellCommand(command);
-	if (ifaceInfo == "")
-	{
-		PCPP_LOG_DEBUG("Error retrieving default gateway address: couldn't get netstat output");
+#elif defined(__APPLE__)
+
+	//route message struct for communication in APPLE device
+	struct BSDRoutingMessage{
+		struct	rt_msghdr header;
+		char	messageSpace[512];
+	};
+
+	struct BSDRoutingMessage routingMessage;
+	char* spacePtr = routingMessage.messageSpace;
+	// It creates a raw socket that can be used for routing-related operations
+	int sockfd = socket(PF_ROUTE, SOCK_RAW, 0);
+	if (sockfd < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't get open routing socket");
+		return ;
+	}
+	memset(reinterpret_cast<char*>(&routingMessage), 0, sizeof(routingMessage));
+	routingMessage.header.rtm_msglen = sizeof(struct rt_msghdr);
+	routingMessage.header.rtm_version = RTM_VERSION;
+	routingMessage.header.rtm_type = RTM_GET;
+	routingMessage.header.rtm_addrs = RTA_DST | RTA_NETMASK | RTA_IFP;
+	routingMessage.header.rtm_flags = RTF_UP | RTF_GATEWAY | RTF_STATIC;
+
+	struct	sockaddr_in so_dst = {0} , so_mask = {0};
+	struct	sockaddr_dl so_ifp = {0};
+
+	size_t len = sizeof(sockaddr_in);
+	bcopy(reinterpret_cast<char*>(&so_dst), spacePtr, len);
+	spacePtr += len;
+	bcopy(reinterpret_cast<char*>(&so_mask), spacePtr, len);
+	spacePtr += len;
+	routingMessage.header.rtm_msglen += 2*len ;
+	len = sizeof(sockaddr_dl);
+	bcopy(reinterpret_cast<char*>(&so_ifp), spacePtr, len);
```

Comment:
I mainly follow the code in the `route.c` source code.I think they want to ensure that in the routing message data space, these structures are laid out in order, so that they can be read when we receive the returning routing message.

I will try the new code see if it works well.

Reply:
I tested it locally without it and it seems to work...

Reply:
removed, make codes look much simpler.